File: algorithms/labs/lab-4/L.java
/**
 * Nariman Safiulin (woofilee)
 * File: L.java
 * Created on: Mar 20, 2016
 */

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class L {
    public static final String PROBLEM_NAME = "salesman";

    public static void main(String[] args) throws Exception {
        Scanner in = new Scanner(new File(PROBLEM_NAME + ".in"));
        PrintWriter out = new PrintWriter(new File(PROBLEM_NAME + ".out"));

        solve(in, out);

        out.close();
        in.close();
    }

    private static void solve(Scanner in, PrintWriter out) throws Exception {
        int n = in.nextInt(), m = in.nextInt();
        int[][] w = new int[n][n], d = new int[1 << n][n];


        for (int i = 0; i < n; ++i)
            Arrays.fill(w[i], Integer.MAX_VALUE);
        for (int i = 0; i < 1 << n; ++i)
            Arrays.fill(d[i], Integer.MAX_VALUE);

        for (int i = 0; i < m; i++) {
            int a = in.nextInt() - 1, b = in.nextInt() - 1;
            w[a][b] = w[b][a] = in.nextInt();
        }

        for (int i = 0; i < n; i++) {
            d[1 << i][i] = 0;
        }

        for (int i = 0; i < 1 << n; i++) {
            for (int j = 0; j < n; j++) {
                if (d[i][j] == Integer.MAX_VALUE)
                    continue;
                for (int k = 0; k < n; k++) {
                    int off = 1 << k;
                    if ((i & off) == 0 && w[j][k] != Integer.MAX_VALUE) {
                        d[i | off][k] = Math.min(d[i | off][k], d[i][j] + w[j][k]);
                    }
                }
            }
        }

        int min = Integer.MAX_VALUE;
        int off = d.length - 1;

        for (int i = 0; i < n; i++) {
            min = Math.min(min, d[off][i]);
        }

        out.println((min == Integer.MAX_VALUE) ? -1 : min);
    }

    static class Scanner {
        BufferedReader br;
        StringTokenizer t;

        Scanner(File file) throws Exception {
            br = new BufferedReader(new FileReader(file));
            t = new StringTokenizer("");
        }

        boolean hasNext() throws Exception {
            while (!t.hasMoreTokens()) {
                String line = br.readLine();
                if (line == null)
                    return false;
                t = new StringTokenizer(line);
            }
            return true;
        }

        String next() throws Exception {
            if (hasNext()) {
                return t.nextToken();
            }
            return null;
        }

        int nextInt() throws Exception {
            return Integer.parseInt(next());
        }

        long nextLong() throws Exception {
            return Long.parseLong(next());
        }

        double nextDouble() throws Exception {
            return Double.parseDouble(next());
        }

        void close() throws Exception {
            br.close();
        }
    }
}

File: labs/lab_4/L.kt
/**
 * Nariman Safiulin (woofilee)
 * File: L.kt
 * Created on: Mar 20, 2016
 */

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.PrintWriter
import java.util.StringTokenizer

private val PROBLEM_NAME = "salesman"

private class Scanner(file: File) {
    val br = BufferedReader(FileReader(file))
    var st = StringTokenizer("")

    fun hasNext(): Boolean {
        while (!st.hasMoreTokens()) {
            val line = br.readLine() ?: return false
            st = StringTokenizer(line)
        }
        return true
    }

    fun next(): String? = if (hasNext()) st.nextToken() else null
    fun nextInt(): Int = Integer.parseInt(next())
    fun nextLong(): Long = java.lang.Long.parseLong(next())
    fun nextDouble(): Double = java.lang.Double.parseDouble(next())
    fun close() = br.close()
}

private fun solve(`in`: Scanner, out: PrintWriter) {
    val n = `in`.nextInt()
    val m = `in`.nextInt()
    val w = Array(n) { Array(n) { Int.MAX_VALUE } }
    val d = Array(1 shl n) { Array(n) { Int.MAX_VALUE } }

    for (i in 0..m - 1) {
        var a = `in`.nextInt() - 1
        var b = `in`.nextInt() - 1
        w[a][b] = `in`.nextInt()
        w[b][a] = w[a][b]
    }
    (0..n - 1).forEach { i -> d[1 shl i][i] = 0 }

    for (i in 0..(1 shl n) - 1) {
        for (j in 0..n - 1) {
            if (d[i][j] == Int.MAX_VALUE) continue
            for (k in 0..n - 1) {
                val off = 1 shl k
                if (i and off == 0 && w[j][k] != Int.MAX_VALUE) {
                    d[i or off][k] = Math.min(d[i or off][k], d[i][j] + w[j][k])
                }
            }
        }
    }

    var min = Int.MAX_VALUE
    val off = d.size - 1
    (0..n - 1).forEach { i -> min = Math.min(min, d[off][i]) }
    out.println(if (min == Int.MAX_VALUE) -1 else min)
}

fun main(args: Array<String>) {
    val `in` = Scanner(File(PROBLEM_NAME + ".in"))
    val out = PrintWriter(File(PROBLEM_NAME + ".out"))

    solve(`in`, out)

    `in`.close()
    out.close()
}
